CtaReactiveMaster/Sources/ViewModels/HomeViewModel.swift

```diff
@@ -0,0 +1,108 @@
+//
+//  HomeViewModel.swift
+//  CtaReactiveMaster
+//
+//  Created by 伊藤凌也 on 2020/12/26.
+//
+
+import RxSwift
+import RxCocoa
+
+protocol HomeViewModelType {
+    var input: HomeViewModelInputs { get }
+    var output: HomeViewModelOutputs { get }
+}
+
+protocol HomeViewModelInputs {
+    func viewDidLoad()
+    func pullToRefresh()
+}
+
+protocol HomeViewModelOutputs {
+    var articles: Driver<[Article]> { get }
+    var showLoading: Signal<Void> { get }
+    var hideLoading: Signal<Void> { get }
+}
+
+struct HomeViewModel: HomeViewModelType, HomeViewModelInputs, HomeViewModelOutputs {
+    private let disposeBag = DisposeBag()
+
+    private let loadingStatus = BehaviorRelay<LoadingStatus>(value: .initial)
+
+    struct Dependency {
+        let repository: NewsRepository
+
+        init(repository: NewsRepository = NewsRepository()) {
+            self.repository = repository
+        }
+    }
+
+    init(dependency: Dependency) {
+        self.articles = articlesRelay.asDriver()
+        self.showLoading = showLoadingRelay.asSignal()
+        self.hideLoading = hideLoadingRelay.asSignal()
+
+        viewDidLoadRelay.asObservable()
+            .map { _ in LoadingStatus.loading }
+            .bind(to: loadingStatus)
+            .disposed(by: disposeBag)
+
+        let fetched = Observable.merge([
+            viewDidLoadRelay.asObservable(),
+            pullToRefreshRelay.asObservable(),
+        ])
+        .flatMap { dependency.repository.fetch().asObservable().materialize() }
+        .share()
+
+        fetched
+            .flatMap { $0.element.map(Observable.just) ?? .empty() }
+            .do(onNext: { [self] _ in
```

Comment:
[imo]
```suggestion
            .do(onNext: { [weak self] _ in
```

今回は、循環参照はしていないからweakキャプチャしなくても問題ないけど、onNextクロージャー内では基本weakキャプチャを行って、レビュワーが循環参照しているかどうかの確認しなくて済むようにするのも良いかなって思うのだけどどうだろう？

Reply:
`struct` なので、そもそも weak で取る必要がない かつ 取れないって感じなのよね

Reply:
あ、structだったのか💡
理解🙆‍♂️

Reply:
質問なんだけどその場合`[self]`を書く必要ってあるの？

Reply:
書けばクロージャ内で `self.` 省略できるという話だったので使ってみてる

あれ、もしかして `[self]` 書かずとも省略できる？

Reply:
りょうちゃんのであってた。ありがとう。